tests/core/celery/tasks/test_generate_inventory_and_hold_reports.py

```diff
@@ -0,0 +1,279 @@
+        pool=licensepool,
+        status=LicenseStatus.available,
+        checkouts_left=checkouts_left,
+        terms_concurrency=terms_concurrency,
+        expires=expiration,
+    )
+    db.license(
+        pool=licensepool,
+        status=LicenseStatus.unavailable,
+        checkouts_left=1,
+        terms_concurrency=1,
+        expires=utc_now(),
+    )
+
+    patron1 = db.patron(library=library)
+    patron2 = db.patron(library=library)
+    patron3 = db.patron(library=library)
+    patron4 = db.patron(library=library)
+
+    # this one should be counted because the end is in the future.
+    hold1, _ = get_one_or_create(
+        db.session,
+        Hold,
+        patron=patron1,
+        license_pool=licensepool,
+        position=1,
+        start=utc_now(),
+        end=utc_now() + timedelta(days=1),
+    )
+
+    # this one should be counted because the end is None
+    hold2, _ = get_one_or_create(
+        db.session,
+        Hold,
+        patron=patron2,
+        license_pool=licensepool,
+        start=utc_now(),
+        end=None,
+    )
+
+    # this hold should be counted b/c the position is > 0
+    hold3, _ = get_one_or_create(
+        db.session,
+        Hold,
+        patron=patron3,
+        license_pool=licensepool,
+        start=utc_now() - timedelta(days=1),
+        end=utc_now() - timedelta(minutes=1),
+        position=1,
+    )
+
+    # this hold should not be counted because the end is neither in the future nor unset and the position is zero
+    hold4, _ = get_one_or_create(
+        db.session,
+        Hold,
+        patron=patron4,
+        license_pool=licensepool,
+        start=utc_now(),
+        end=utc_now() - timedelta(minutes=1),
+        position=0,
+    )
+
+    shared_patrons_in_hold_queue = 4
+    licensepool.patrons_in_hold_queue = shared_patrons_in_hold_queue
+
+    assert library.id
+
+    job = GenerateInventoryAndHoldsReportsJob(
+        mock_session_maker, library.id, email_address=email
+    )
+
+    send_email_mock = create_autospec(job.services.email.container.send_email)
+    # for testing, don't delete the files associated with the attachments so we can read them after the script
+    # runs
+    GenerateInventoryAndHoldsReportsJob._delete_attachments = False
+    job.services.email.container.send_email = send_email_mock
+    job.run()
+    send_email_mock.assert_called_once()
+    args, kwargs = send_email_mock.call_args
+    assert kwargs["receivers"] == [email]
+    assert "Inventory and Holds Reports" in kwargs["subject"]
+    attachments: dict = kwargs["attachments"]
+
+    assert len(attachments) == 2
+    inventory_report_key = [x for x in attachments.keys() if "inventory" in x][0]
+    assert inventory_report_key
+    assert "test_library" in inventory_report_key
+    inventory_report_value = attachments[inventory_report_key]
+    assert inventory_report_value
+    inventory_report_reader = csv.DictReader(open(inventory_report_value))
+    inventory_row_count = 0
```

Comment:
Instead of doing `inventory_row_count` like this, it would be nice to test it before going into the loop, so that we know if we get the wrong number of results before. You could do something like this I think:

```python
inventory_report_csv = list(csv.DictReader(open(inventory_report_value)))

assert len(inventory_report_csv) == 1

for row in inventory_report_csv:
   ...
```